File: gcj_2017/QR_A.hs
-- Google Code Jam 2017 Qualification Round
-- Problem A. Oversized Pancake Flipper

import Control.Monad
import Data.Char
import Text.Printf

readInt = (read <$> getLine) :: IO Int

rev :: String -> String
rev = map (\c -> if c == '+' then '-'; else '+')

f :: String -> Int -> Int -> Int
f s k t
  | (length s) == 0 = t
  | (head s) == '+' = f (tail s) k t
  | (length s) < k = -1
  | otherwise = f ((rev (take k s)) ++ (drop k s)) k (1 + t)

solve :: String -> Int -> String
solve s k = do
  let ans = f s k 0
  if ans < 0 then "IMPOSSIBLE" else show ans

main :: IO ()
main = do
  t <- readInt
  forM_ [1..t] $ \i -> do
    [s, k] <- words <$> getLine
    putStrLn $ "Case #" ++ (show i) ++ ": " ++ (solve s (read k :: Int))

File: gcj_2017/QR_A.cpp
// Google Code Jam 2017 Qualification Round
// Problem A. Oversized Pancake Flipper

#include <iostream>
#include <sstream>

using namespace std;

int solve(string s, int k) {
	if (s.empty()) {
		return 0;
	}
	if (s[0] == '+') {
		return solve(s.substr(1), k);
	}
	if (k > s.length()) {
		return -1;
	}
	for (int i = 0; i < k; ++i) {
		s[i] = s[i] == '+' ? '-' : '+';
	}
	int r = solve(s.substr(1), k);
	return r < 0 ? -1 : 1 + r;
}

int main(int argc, char *argv[]) {
	int T;
	cin >> T;
	for (int t = 1; t <= T; ++t) {
		string s;
		int k;
		cin >> s >> k;
		int ans = solve(s, k);
		if (ans < 0) {
			cout << "Case #" << t << ": IMPOSSIBLE" << endl;
		} else {
			cout << "Case #" << t << ": " << ans << endl;
		}
	}
	return 0;
}
